overlord/ifacestate/apparmorprompting/prompting.go

```diff
@@ -0,0 +1,578 @@
+		return nil, err
+	}
+
+	// Check that constraints matches original requested path.
+	// AppArmor is responsible for pre-vetting that all paths which appear
+	// in requests from the kernel are allowed by the appropriate
+	// interfaces, so we do not assert anything else particular about the
+	// constraints, such as check that the path pattern does not match
+	// any paths not granted by the interface.
+	// TODO: Should this be reconsidered?
+	matches, err := constraints.Match(prompt.Constraints.Path())
+	if err != nil {
+		return nil, err
+	}
+	if !matches {
+		return nil, fmt.Errorf("constraints in reply do not match original request: '%v' does not match '%v'; please try again", constraints, prompt.Constraints)
+	}
+
+	// XXX: do we want to allow only replying to a select subset of permissions, and
+	// auto-deny the rest?
+	contained := constraints.ContainPermissions(prompt.Constraints.RemainingPermissions())
+	if !contained {
+		return nil, fmt.Errorf("replied permissions do not include all requested permissions: requested %v, replied %v; please try again", prompt.Constraints.RemainingPermissions(), constraints.Permissions)
+	}
+
+	// It is important that a lock is held while checking for conflicts with
+	// other rules so that if the rule is eventually removed due to an error,
+	// no prompts can have been matched against it in the meantime.
+	var newRule *requestrules.Rule
+	if lifespan != prompting.LifespanSingle {
+		// Check that adding the rule doesn't conflict with other rules
+		newRule, err = m.rules.AddRule(userID, prompt.Snap, prompt.Interface, constraints, outcome, lifespan, duration)
+		if err != nil {
+			// Rule conflicts with existing rule (at least one identical pattern
+			// variant and permission). This should be considered a bad reply,
+			// since the user should only be prompted for permissions and paths
+			// which are not already covered.
+			return nil, err
+		}
+
+		defer func() {
+			if retErr != nil || lifespan == prompting.LifespanSingle {
+				m.rules.RemoveRule(userID, newRule.ID)
+			}
+		}()
+	}
+
+	prompt, retErr = m.prompts.Reply(userID, promptID, outcome)
+	if retErr != nil {
+		// Error should not occur unless the listener has closed
+		return nil, retErr
+	}
+
+	if lifespan == prompting.LifespanSingle {
+		return []prompting.IDType{}, nil
+	}
+
+	// Apply new rule to outstanding prompts.
+	satisfiedPromptIDs, err = m.applyRuleToOutstandingPrompts(newRule)
+	if err != nil {
+		// Should not occur, as outcome and constraints have already been
+		// validated. However, it's possible an error could occur if the prompt
+		// DB was already closed. This should be an internal error. However, we
+		// can't un-send any reply which might have been sent, and this should
+		// only fail if the prompting system is shutting down, so don't actually
+		// return an error.
+
+		// XXX: this log leaks information about internal activity
+		logger.Noticef("error when handling new rule as a result of reply: %v", err)
```

Comment:
as we are doing this in all places we are calling  m.applyRuleToOutstandingPrompts maybe we should move it and have it not return an error?